platooning_strategic_IHP/src/platoon_strategic_ihp.cpp

```diff
@@ -298,7 +303,105 @@ namespace platoon_strategic_ihp
+         * 
+         * @return
+         * target_lanelet_id: The lanelet id of the target lane.
+         */
+
+        std::vector<lanelet::ConstLanelet> lanelets_in_path = wm_->getLaneletsBetween(start_downtrack, end_downtrack, false, false);
+        std::vector<double> downtracks,crosstracks;
+        std::vector<lanelet::ConstLanelet> lanelets_samelane,lanelets_sidelane;
+        lanelet::ConstLanelet target_lanelet;
+        int target_lanelet_id;
+        if (!lanelets_in_path.empty())
+        {
+            for (int i = 0; i < lanelets_in_path.size(); ++i)
+            {
+                lanelet::BasicPoint2d p1=lanelets_in_path[i].centerline2d().basicLineString().back();
+                double downtrack=wm_->routeTrackPos(p1).downtrack;
+                double crosstrack=wm_->routeTrackPos(p1).crosstrack;
+                if (abs(crosstrack)>1)//1 is a threshold, the assumption for this condition is that there are two lanes on the road
+                {
+                    lanelets_sidelane.push_back(lanelets_in_path[i]);
+                    downtracks.push_back(downtrack);
+                    crosstracks.push_back(crosstrack);
+                }else
+                {
+                    lanelets_samelane.push_back(lanelets_in_path[i]);
+                }
+            }
+            if (!downtracks.empty())
+            {
+                int maxPosition = max_element(downtracks.begin(),downtracks.end()) - downtracks.begin();
+                target_lanelet=lanelets_sidelane[maxPosition];
+                target_lanelet_id=target_lanelet.id();
+            }
+        }
+        return target_lanelet_id;
+    } 
+
+    // Check if the platoon close to the platoon for cut-in join.
+    bool PlatoonStrategicIHPPlugin::isVehicleNextLane(double frontVehicleDtd, double rearVehicleDtd, double frontVehicleCtd)
+    {
+        /**
+         * @brief Function to determine if the host vehicle is close to the target platoon
+         * 
+         * @param 
+         *   (double) front downtrack: the downtrack of the target platoon leader.
+         *   (double) rear downtrack: the downtrack of the last target platoon member.
+         *   (double) front crosstrack: the crosstrack of the target platoon leader. 
+         * @return
+         *   (bool): if the target leader behind
+         */
+
+       double currentDtd = current_downtrack_;
+       double currentCtd = current_crosstrack_;
+       // lateral error for two lanes 
+       double two_lane_cross_error = 2*config_.maxCrosstrackError + 3.5; // todo: 3.5m is lane width, consider read it from map
+       // todo: use maxgap (35m) as longitudinal threshold, consider change 
+       bool longitudinalCheck = ((currentDtd >= rearVehicleDtd - maxAllowedJoinGap_) || (currentDtd <= frontVehicleDtd + 10));
+       bool lateralCheck = ((currentCtd >= frontVehicleCtd - two_lane_cross_error) || (currentCtd <= frontVehicleCtd + two_lane_cross_error));
```

Comment:
Clever test!  I suppose it may result in false sometimes if vehicles are both to the far sides of their respective lanes AND there is some measurement error.  Worth thinking about just a little bit more error allowance for that, or accepting that sometimes it will just fail.  At least you have logging for a failure.